src/metabase/db/custom_migrations/metrics_v2.clj

```diff
@@ -0,0 +1,158 @@
+(defn- metric-ref->id
+  "If `expr` is a metric reference expression, return the ID of the referenced metric.
+  Returns nil otherwise."
+  [expr]
+  (when (and (vector? expr)
+             (= (u/lower-case-en (get expr 0)) "metric"))
+    (let [id (get expr 1)]
+      ;; GA metrics with string references are ignored
+      (when (int? id)
+        id))))
+
+(defn- replace-metric-refs
+  "Replaces the IDs in metric references contained in `expr` with the :id properties
+  of the corresponding entities in `id->entity`.
+  `expr` is an MBQL aggregation expression or a part of it.
+  The mapping is from metric IDs to v2 metric cards."
+  [expr id->entity]
+  (if-not (vector? expr)
+    expr
+    (if-let [id (metric-ref->id expr)]
+      (assoc expr 1 (-> id id->entity :id))
+      (mapv #(replace-metric-refs % id->entity) expr))))
+
+(defn- rewrite-metric-consuming-query
+  [query metric-id->metric-card]
+  (if (contains? query :source-query)
+    (update query :source-query rewrite-metric-consuming-query metric-id->metric-card)
+    (let [aggregation (:aggregation query)
+          rewritten (replace-metric-refs aggregation metric-id->metric-card)]
+      (cond-> query
+        (not= rewritten aggregation) (assoc :aggregation rewritten)))))
+
+(defn- rewrite-metric-consuming-card
+  "Rewrite the question `card` replacing references to v1 metrics with references
+  to the corresponding v2 metric question as specified by the mapping `metric-id->metric-card`."
+  [card metric-id->metric-card]
+  (let [dataset-query (json/parse-string (:dataset_query card) true)
+        inner-query (:query dataset-query)
+        rewritten (rewrite-metric-consuming-query inner-query metric-id->metric-card)]
+    (when (not= rewritten inner-query)
+      (-> card
+          (assoc :dataset_query (-> dataset-query
+                                    (assoc :query rewritten)
+                                    json/generate-string))
+          (dissoc :updated_at)))))
+
+(defn migrate-up
+  "Migrate metrics and the cards consuming them to metrics v2. This involves
+  1. creating the migration collection,
+  2. creating a v2 metric card for each v1 metric in the migration collection,
+  3. making a backup copy of the dataset_query of each metric consuming card,
+  4. rewriting the metric consuming cards to use the migrated v2 metric cards."
+  []
+  (when (t2/exists? :metric)
+    (let [metric-v2-coll (get-metric-migration-collection {:create? true})
+          metric-id->metric-card (into {}
+                                       (map (juxt :id #(create-metric-v2 % metric-v2-coll)))
+                                       (t2/query {:select [:m.* [:t.db_id :database_id]]
+                                                  :from [[:metric :m]]
+                                                  :left-join [[:metabase_table :t] [:= :t.id :m.table_id]]}))
+          metric-consuming-cards (t2/select :report_card {:where [:like [:lower :dataset_query] "%[\"metric\",%"]})]
```

Comment:
```suggestion
          metric-consuming-cards (t2/select [:report_card :id :dataset_query] {:where [:like [:lower :dataset_query] "%[\"metric\",%"]})]
```

You can make this a lot more efficient by only fetching the columns you actually need, like this. This does `SELECT id, dataset_query` instead of `SELECT *`, which will make the results a **lot** smaller.